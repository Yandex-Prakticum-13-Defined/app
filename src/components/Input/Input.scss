.input {
  padding: 5px 25px;
  border: none;
  border-bottom: 3px solid black;
  border-radius: 20px;

  &__error {
    color: red;
    margin-left: 10px;
    font-size: 13px;
    height: 33px;
  }
}
