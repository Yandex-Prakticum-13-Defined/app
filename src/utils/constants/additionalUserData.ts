export const additionalUserData = {
  second_name: 'Second',
  display_name: 'Display',
  phone: '+7-000-000-00-00'
};
