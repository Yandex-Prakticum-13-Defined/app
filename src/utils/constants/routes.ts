export enum ERoutes {
  'START' = '/',
  'REGISTER' = '/register',
  'LOGIN' = '/login',
  'LOGOUT' = '/logout',
  'GAME' = '/game',
  'LEADERBOARD' = '/leaderboard',
  'PROFILE' = '/profile',
  'FORUM' = '/forum',
  'ERROR_500' = '/500',
  'OFFLINE' = '/offline',
  'FALLBACK' = '*'
}

export enum EDBRoutes {
  'TOPICS' = '/topics',
  'MESSAGES' = '/messages',
  'TOPIC' = '/topic',
  'MESSAGE' = '/message'
}
