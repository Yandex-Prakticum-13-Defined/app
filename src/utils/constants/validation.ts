export const VALIDATION = {
  required: 'Поле не может быть пустым',
  name: {
    value: /^[A-ZА-ЯЁ]+[A_Za-zА-Яа-яЁё-]+/,
    message: 'Латиница или кириллица, первая буква заглавная, '
      + 'без пробелов и цифр, нет спецсимволов (допустим только дефис)'
  },
  login: {
    value: /[a-zA-Z0-9_-]*[a-zA-Z_-][a-zA-Z0-9_-]*/,
    message: 'Латиница, может содержать цифры, '
      + 'но не состоять из них, допустимы дефис и нижнее подчёркивание'
  },
  email: {
    value: /^[A-Za-z0-9,.-]{1,}[@]([A-Za-z0-9,.-]{1,}[.][A-Za-z]{1,}){1,2}$/,
    message: 'Латиница, может включать цифры и '
      + 'спецсимволы вроде дефиса, обязательна @ и точка после,'
      + ' но перед точкой обязательно должны быть буквы'
  },
  password: {
    value: /(?=.*[A-Z])(?=.*[0-9]).*/,
    message: 'От 8 до 40 символов, обязательно хотя бы одна заглавная буква и цифра'
  },
  minLength_8: {
    value: 8,
    message: 'Минимальное количество символов: 8'
  },
  minLength_3: {
    value: 3,
    message: 'Минимальное количество символов: 3'
  },
  maxLength_20: {
    value: 20,
    message: 'Максимальное количество символов: 20'
  }
};
