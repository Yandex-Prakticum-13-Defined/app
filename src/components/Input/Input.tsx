import React from 'react';
import './Input.scss';
import { useController } from 'react-hook-form';

export interface IInput {
  name: string;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  control: any;
  value?: string;
  rules?: Record<string, unknown>;
  type: string;
  placeholder: string;
}

// eslint-disable-next-line react/display-name
export const Input = ({
  name,
  control,
  value,
  rules,
  ...rest
}: IInput) => {
  const {
    field,
    formState: { errors }
  } = useController({
    name,
    control,
    defaultValue: value ?? '',
    rules
  });

  return (
    <>
      {/* eslint-disable-next-line react/jsx-props-no-spreading */}
      <input className='input' {...field} {...rest}/>
      {errors && <span className='input__error'>{errors[name]?.message}</span>}
    </>
  );
};
