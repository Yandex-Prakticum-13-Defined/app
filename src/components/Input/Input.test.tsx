import React, { FC } from 'react';
import renderer from 'react-test-renderer';
import { useForm } from 'react-hook-form';
import { act, render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { VALIDATION } from '../../utils/constants/validation';
import { Input } from './Input';

const TestInputWrapper: FC = () => {
  const { control } = useForm({ mode: 'onChange' });

  return (
    <Input
      name='test'
      control={control}
      value='Значение по умолчанию'
      rules={{
        required: VALIDATION.required,
        minLength: VALIDATION.minLength_3,
        maxLength: VALIDATION.maxLength_20,
        pattern: VALIDATION.login
      }}
      type='text'
      placeholder='Введите значение'
    />
  );
};

describe('Тестируем Input.tsx', () => {
  it('проверяем снимок', () => {
    const inputComponent = renderer.create(
      <TestInputWrapper/>
    ).toJSON();
    expect(inputComponent).toMatchSnapshot();
  });

  it('передаётся value', () => {
    render(<TestInputWrapper/>);
    expect(screen.getByDisplayValue('Значение по умолчанию')).toBeInTheDocument();
  });

  it('передаётся placeholder', () => {
    render(<TestInputWrapper/>);
    expect(screen.getByPlaceholderText('Введите значение')).toBeInTheDocument();
  });

  it('валидируется required', async () => {
    render(<TestInputWrapper/>);
    const user = userEvent.setup();
    await act(async () => {
      await user.clear(screen.getByPlaceholderText('Введите значение'));
    });
    expect(screen.getByText(VALIDATION.required)).toBeInTheDocument();
  });

  it('валидируется minLength', async () => {
    render(<TestInputWrapper/>);
    const user = userEvent.setup();
    await act(async () => {
      await user.clear(screen.getByPlaceholderText('Введите значение'));
    });
    await act(async () => {
      await user.type(screen.getByPlaceholderText('Введите значение'), '1');
    });
    expect(screen.getByText(VALIDATION.minLength_3.message)).toBeInTheDocument();
  });

  it('валидируется maxLength', async () => {
    render(<TestInputWrapper/>);
    const user = userEvent.setup();
    await act(async () => {
      await user.clear(screen.getByPlaceholderText('Введите значение'));
    });
    await act(async () => {
      await user.type(screen.getByPlaceholderText('Введите значение'), 'Здесь явно больше 20 символов');
    });
    expect(screen.getByText(VALIDATION.maxLength_20.message)).toBeInTheDocument();
  });

  it('валидируется pattern', async () => {
    render(<TestInputWrapper/>);
    const user = userEvent.setup();
    await act(async () => {
      await user.clear(screen.getByPlaceholderText('Введите значение'));
    });
    await act(async () => {
      await user.type(screen.getByPlaceholderText('Введите значение'), '12345');
    });
    expect(screen.getByText(VALIDATION.login.message)).toBeInTheDocument();
  });
});
