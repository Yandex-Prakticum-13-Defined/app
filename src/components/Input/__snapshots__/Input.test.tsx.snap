// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Тестируем Input.tsx проверяем снимок 1`] = `
Array [
  <input
    className="input"
    name="test"
    onBlur={[Function]}
    onChange={[Function]}
    placeholder="Введите значение"
    type="text"
    value="Значение по умолчанию"
  />,
  <span
    className="input__error"
  />,
]
`;
